### PR TITLE
Move some common test-utils to their own crate

### PR DESCRIPTION
**Pull request by nicklan**
## What changes are proposed in this pull request?

Move some useful common utils for building an in-memory delta table into their own crate.

I want to start adding tests for actually reading data to the `ffi` crate. To do so we need to use an `InMemory` object store, because all i/o in tokio fails in miri due to it using epoll, which miri doesn't support. So we'll need to build in-memory tables and use the `memory:///` url to get a memory backed `ObjectStore`. 

We already do this in `read.rs` in kernel, but there's no way to get at that code inside the ffi crate. So this moves a bunch of utilities for this sort of thing into its own "test-utils" crate and uses it in read.rs.

I've also cleaned up and renamed things a bit. Most of the code remains unchanged.

Follow-up PR(s) will use this in the ffi crate to test more.
## How was this change tested?

Existing tests pass

### CODE REVIEW

**Review by zachschuermann — APPROVED** [review]
nit: is `test-utils` or `test_utils` more idiomatic I don't remember? Also does naming not matter since this won't ever make it to crates.io?

everything looks great thanks for the cleanup - I think one thing we could do better (that i've been meaning to get to soon) is make the raw JSON (like for `const METADATA`) using `json!` macro from serde just for readability. just wanted to mention that - but happy to do that later!!

**Comment by zachschuermann** [comment]
weird semver fails due to
```
failed to parse /home/runner/work/delta-kernel-rs/delta-kernel-rs/target/semver-checks/git-9861282e77ba1e3ef3410780d490feab52e90b18/83da59f23e419fe81cc19b13b6750f4c634b28bb/Cargo.toml: no `package` table,
```

**Comment by nicklan** [comment]
> weird semver fails due to
> 
> ```
> failed to parse /home/runner/work/delta-kernel-rs/delta-kernel-rs/target/semver-checks/git-9861282e77ba1e3ef3410780d490feab52e90b18/83da59f23e419fe81cc19b13b6750f4c634b28bb/Cargo.toml: no `package` table,
> ```

Ahh, it's because it's trying to compare against `main` which doesn't have `test_utils`. Will only be an issue for this PR

**Comment by nicklan** [comment]
> nit: is `test-utils` or `test_utils` more idiomatic I don't remember? Also does naming not matter since this won't ever make it to crates.io?

Yeah, doesn't really matter, but we've gone `camel_case` for all the other crates we have, so I've updated it to make it consistent.

File: test-utils/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+use delta_kernel::engine::arrow_data::ArrowEngineData;
+use delta_kernel::EngineData;
+use itertools::Itertools;
+use object_store::{path::Path, ObjectStore};
+use parquet::arrow::arrow_writer::ArrowWriter;
+use parquet::file::properties::WriterProperties;
+
+/// A common useful initial metadata and protocol. Also includes a single commitInfo
+pub const METADATA: &str = r#"{"commitInfo":{"timestamp":1587968586154,"operation":"WRITE","operationParameters":{"mode":"ErrorIfExists","partitionBy":"[]"},"isBlindAppend":true}}
+{"protocol":{"minReaderVersion":1,"minWriterVersion":2}}
```
Comment (OussamaSaoudi-db) [inline test-utils/src/lib.rs:16]: Would it be possible to make this a `Metadata` struct that we convert to string? I feel like that's easier to read.

Comment (nicklan) [reply test-utils/src/lib.rs:16]: Yeah, there's lots of cleanup we can do. I'm preferring to just mostly move the code to its own crate (and a couple of minor cleanups) for now, and we can take on making this "better" in the future.

File: test-utils/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    ])
+}
+
+/// put a commit file into the specified object store.
+pub async fn add_commit(
+    store: &dyn ObjectStore,
+    version: u64,
+    data: String,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let filename = format!("_delta_log/{:0>20}.json", version);
```
Comment (OussamaSaoudi-db) [inline test-utils/src/lib.rs:91]: I've seen this line of code a few times in the codebase. Might be worth making it into a common util `delta_path_from_version` or smth.

File: test-utils/src/lib.rs (hunk 1 above)
Comment (OussamaSaoudi-db) [reply test-utils/src/lib.rs:16]: sounds good! 👍

**Review by OussamaSaoudi-db — APPROVED** [review]
Nothing else besides the potential future work.  LGTM

File: test-utils/src/lib.rs (hunk 2 above)
Comment (nicklan) [reply test-utils/src/lib.rs:91]: > I've seen this line of code a few times in the codebase. Might be worth making it into a common util `delta_path_from_version` or smth.

Good call, done

**Pull request merged**